pkg/models/terminal/terminal.go

```diff
@@ -140,15 +151,135 @@ func (t TerminalSession) Close(status uint32, reason string) {
+}
+
+func NewNodeTerminaler(nodename string, options *Options, client kubernetes.Interface) (*NodeTerminaler, error) {
+
+	n := &NodeTerminaler{
+		Namespace:     "kubesphere-controls-system",
+		ContainerName: "nsenter",
+		Nodename:      nodename,
+		PodName:       nodename + "-shell-access",
+		Shell:         "sh",
+		Privileged:    true,
+		Config:        options,
+		client:        client,
+	}
+
+	node, err := n.client.CoreV1().Nodes().Get(context.Background(), n.Nodename, metav1.GetOptions{})
+
+	if err != nil {
+		return n, fmt.Errorf("node cannot exist. nodename:%s, err: %v", n.Nodename, err)
+	}
+
+	flag := false
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == v1.NodeReady && condition.Status == v1.ConditionTrue {
+			flag = true
+		}
+	}
+	if !flag {
+		return n, fmt.Errorf("node status error. node: %s", n.Nodename)
+	}
+
+	mu.Lock()
+	if _, ok := NodeSessionCounter[nodename]; !ok {
+		var index int64 = 0
+		NodeSessionCounter[nodename] = &index
+	}
+	mu.Unlock()
+
+	return n, nil
+}
+
+func (n *NodeTerminaler) getNSEnterPod() (*v1.Pod, error) {
+	pod, err := n.client.CoreV1().Pods(n.Namespace).Get(context.Background(), n.PodName, metav1.GetOptions{})
+
+	if err != nil || (pod.Status.Phase != v1.PodRunning && pod.Status.Phase != v1.PodPending) {
+		//pod has timed out, but has not been cleaned up
+		if pod.Status.Phase == v1.PodSucceeded || pod.Status.Phase == v1.PodFailed {
+			err := n.client.CoreV1().Pods(n.Namespace).Delete(context.Background(), n.PodName, metav1.DeleteOptions{})
+			if err != nil {
+				return pod, err
+			}
+		}
+
+		var p = &v1.Pod{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "pod",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: n.PodName,
+			},
+			Spec: v1.PodSpec{
+				NodeName:      n.Nodename,
+				HostPID:       true,
+				HostNetwork:   true,
+				RestartPolicy: v1.RestartPolicyNever,
+				Containers: []v1.Container{
+					{
+						Name:  n.ContainerName,
+						Image: n.Config.Image,
+						Command: []string{
+							"nsenter", "-m", "-u", "-i", "-n", "-p", "-t", "1",
+						},
+						Stdin: true,
+						TTY:   true,
+						SecurityContext: &v1.SecurityContext{
+							Privileged: &n.Privileged,
+						},
+					},
+				},
+			},
+		}
+
+		if n.Config.Timeout == 0 {
+			p.Spec.Containers[0].Args = []string{"tail", "-f", "/dev/null"}
+		} else {
+			p.Spec.Containers[0].Args = []string{"sleep", strconv.Itoa(n.Config.Timeout)}
+		}
+
+		pod, err = n.client.CoreV1().Pods(n.Namespace).Create(context.Background(), p, metav1.CreateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("create pod failed on %s node: %v", n.Nodename, err)
+		}
+	}
+
+	return pod, nil
 }
 
-func NewTerminaler(client kubernetes.Interface, config *rest.Config) Interface {
-	return &terminaler{client: client, config: config}
+func (n NodeTerminaler) CleanUpNSEnterPod() {
+	atomic.AddInt64(NodeSessionCounter[n.Nodename], -1)
```

Comment:
If at one second before,  another session comes in and adds mutex lock in its session, do many sessions be blocked to wait for that session? So do we need to weaken the lock to write/read lock?  Or have you tested the sync.Map?

Reply:
Only a simple variable assignment is made in the lock, which will not take too long. I think there shouldn't cause a long wait

Reply:
OK. IMO, I think sync.Map will improve the read performance. You can have a try.

Reply:
> OK. IMO, I think sync.Map will improve the read performance. You can have a try.

Sync.map is thread-safe, with constant time complexity for reads, inserts, and deletions.  You can remove lock here.

Reply:
OK.I'll modify it